File: src/voicevox_client.py
"""
VOICEVOX Engine API クライアント（栗田まろん）

前提:
  - VOICEVOX Engine がローカル http://localhost:50021 で起動中
  - 環境変数 VOICEVOX_HOST で変更可能

機能:
  1. /speakers から「栗田まろん」の speaker_id を動的取得
  2. 全文を句点（。！？）で分割
  3. 各文を /audio_query → /synthesis で合成
  4. 文間に 0.3 秒の無音を挿入して 1 つの WAV に結合

使い方:
  python -m src.voicevox_client --input output/reels/2026-06-01/script.json
  python -m src.voicevox_client --text "テスト音声" --output test.wav
"""

from __future__ import annotations

import argparse
import io
import json
import os
import re
import sys
import wave
from pathlib import Path

import requests


ROOT = Path(__file__).resolve().parents[1]
VOICEVOX_HOST = os.environ.get("VOICEVOX_HOST", "http://localhost:50021")
DEFAULT_SPEAKER_NAME = "青山龍星"
DEFAULT_STYLE_NAME = "喜び"   # None なら最初のスタイルを使う

# 読み上げパラメータ
# - intonationScale を 1.0（VOICEVOX 既定）に戻すと、語尾の上下動が自然になる
# - 1.0 以上にすると抑揚が誇張されて語尾が下がらず不自然になりやすい
DEFAULT_SPEED = 1.05       # リール向けに少しだけ速め
DEFAULT_PITCH = 0.0
DEFAULT_INTONATION = 1.0   # 既定。語尾のイントネーションを自然に保つ
SILENCE_BETWEEN_SEC = 0.25 # 文間の無音（短めに）

# === 語尾下げ調整（B案）===
# 各アクセント句の末尾モーラの pitch を強制的に下げて、語尾の上がりを抑える。
# 疑問文（is_interrogative=True）は対象外（上がるべきため）。
# 値は経験的な係数。0.88 = 末尾を 12% 下げる、0.94 = 2つ前を 6% 下げる。
LAST_MORA_PITCH_FACTOR = 0.88   # 末尾モーラ
SECOND_LAST_MORA_PITCH_FACTOR = 0.94   # 2つ前のモーラ（自然なグライド用）

# Windows コンソール UTF-8
if hasattr(sys.stdout, "reconfigure"):
    try:
        sys.stdout.reconfigure(encoding="utf-8", errors="replace")
        sys.stderr.reconfigure(encoding="utf-8", errors="replace")
    except Exception:
        pass


class VOICEVOXError(Exception):
    pass


def check_engine() -> bool:
    """VOICEVOX Engine が稼働中かどうかを確認。"""
    try:
        r = requests.get(f"{VOICEVOX_HOST}/version", timeout=3)
        return r.status_code == 200
    except requests.RequestException:
        return False


def find_speaker_id(
    speaker_name: str = DEFAULT_SPEAKER_NAME,
    style_name: str | None = DEFAULT_STYLE_NAME,
) -> int:
    """speaker name（+ style name）から style_id を動的取得。

    - style_name 指定時はそのスタイルの id を返す
    - style_name=None の場合は最初のスタイル
    - スタイルが見つからない場合は利用可能なスタイル一覧を表示
    """
    r = requests.get(f"{VOICEVOX_HOST}/speakers", timeout=30)
    r.raise_for_status()
    speakers = r.json()
    for sp in speakers:
        if sp.get("name") == speaker_name:
            styles = sp.get("styles", [])
            if not styles:
                raise VOICEVOXError(f"スピーカー '{speaker_name}' にスタイルがありません")
            if style_name:
                for st in styles:
                    if st.get("name") == style_name:
                        return st.get("id")
                avail_styles = [st.get("name") for st in styles]
                raise VOICEVOXError(
                    f"スピーカー '{speaker_name}' にスタイル '{style_name}' がありません。\n"
                    f"利用可能なスタイル: {avail_styles}"
                )
            return styles[0].get("id")
    available = sorted({sp.get("name", "") for sp in speakers})
    raise VOICEVOXError(
        f"スピーカー '{speaker_name}' が見つかりません。\n"
        f"利用可能なスピーカー一覧 ({len(available)}名):\n"
        + "\n".join(f"  - {n}" for n in available)
    )


def _audio_query(text: str, speaker_id: int) -> dict:
    r = requests.post(
        f"{VOICEVOX_HOST}/audio_query",
        params={"text": text, "speaker": speaker_id},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def _synthesis(query: dict, speaker_id: int) -> bytes:
    r = requests.post(
        f"{VOICEVOX_HOST}/synthesis",
        params={"speaker": speaker_id},
        json=query,
        timeout=120,
    )
    r.raise_for_status()
    return r.content  # WAV bytes


def _flatten_phrase_endings(
    query: dict,
    last_factor: float = LAST_MORA_PITCH_FACTOR,
    second_last_factor: float = SECOND_LAST_MORA_PITCH_FACTOR,
) -> dict:
    """文末（最後のアクセント句）の末尾モーラの pitch を下げて、語尾の上がりを抑える。

    重要: 文中のアクセント句（助詞や読点前など）には触れない。
    そうしないと文のリズムが壊れて「外国人が日本語を話している」ような不自然な調子になる。

    - 「？」も含めてすべて下げる（VOICEVOX の ？ 上がりは過剰なため）
    - pitch=0 のモーラ（無声子音など）はスキップ
    - 末尾モーラと 2 つ前のモーラを段階的に下げる（自然なグライド）
    - query は in-place で書き換え、参照のため return もする
    """
    phrases = query.get("accent_phrases") or []
    if not phrases:
        return query

    # 文末のアクセント句のみ対象
    last_phrase = phrases[-1]
    moras = last_phrase.get("moras") or []
    if not moras:
        return query

    # 末尾モーラ
    last = moras[-1]
    if last.get("pitch", 0) > 0:
        last["pitch"] = round(last["pitch"] * last_factor, 3)

    # 2 つ前のモーラ（グライドを自然にするため、軽めに下げる）
    if len(moras) >= 2:
        second = moras[-2]
        if second.get("pitch", 0) > 0:
            second["pitch"] = round(second["pitch"] * second_last_factor, 3)

    return query


def synthesize_one(
    text: str,
    speaker_id: int,
    speed: float = DEFAULT_SPEED,
    pitch: float = DEFAULT_PITCH,
    intonation: float = DEFAULT_INTONATION,
    flatten_endings: bool = True,
) -> bytes:
    """1 文を合成して WAV bytes を返す。

    flatten_endings=True の場合、非疑問句の末尾モーラを下げて
    語尾の上がりを抑制する（B案）。
    """
    query = _audio_query(text, speaker_id)
    query["speedScale"] = speed
    query["pitchScale"] = pitch
    query["intonationScale"] = intonation
    if flatten_endings:
        _flatten_phrase_endings(query)
    return _synthesis(query, speaker_id)


def split_into_sentences(text: str) -> list[str]:
    """。！？ で分割（区切り文字を含めて残す）。"""
    parts = re.split(r"(?<=[。！？])", text)
    return [p.strip() for p in parts if p.strip()]


def synthesize_full(
    full_script: str,
    speaker_id: int,
    silence_sec: float = SILENCE_BETWEEN_SEC,
    flatten_endings: bool = True,
) -> tuple[bytes, list[dict]]:
    """全文を句点で分割→個別合成→文間無音を挟んで結合した WAV を返す。

    flatten_endings=True (既定) で各句末尾のピッチを下げて語尾上がりを抑制する。

    返り値:
      (wav_bytes, timings)
      timings = [{"text": "...", "start": 0.00, "end": 1.83}, ...]
        各文の実 WAV 長から計算した字幕タイミング（秒）。
    """
    sentences = split_into_sentences(full_script)
    if not sentences:
        raise VOICEVOXError("台本が空です")

    pcm_chunks: list[bytes] = []
    sentence_durations: list[float] = []
    sample_rate: int | None = None
    sample_width: int | None = None
    n_channels: int | None = None

    for sent in sentences:
        wav_bytes = synthesize_one(sent, speaker_id, flatten_endings=flatten_endings)
        with wave.open(io.BytesIO(wav_bytes), "rb") as w:
            if sample_rate is None:
                sample_rate = w.getframerate()
                sample_width = w.getsampwidth()
                n_channels = w.getnchannels()
            pcm = w.readframes(w.getnframes())
            pcm_chunks.append(pcm)
            sentence_durations.append(w.getnframes() / w.getframerate())

    # 無音 PCM
    silence_frames = int(sample_rate * silence_sec)
    silence_pcm = b"\x00" * (silence_frames * sample_width * n_channels)

    # 結合 + タイミング構築（音声の実際の連結順序と完全一致させる）
    combined = b""
    timings: list[dict] = []
    cursor = 0.0
    for i, (sent, pcm, dur) in enumerate(zip(sentences, pcm_chunks, sentence_durations)):
        timings.append({"text": sent, "start": round(cursor, 3), "end": round(cursor + dur, 3)})
        combined += pcm
        cursor += dur
        if i < len(sentences) - 1:
            combined += silence_pcm
            cursor += silence_sec

    out_buf = io.BytesIO()
    with wave.open(out_buf, "wb") as w:
        w.setnchannels(n_channels)
        w.setsampwidth(sample_width)
        w.setframerate(sample_rate)
        w.writeframes(combined)
    return out_buf.getvalue(), timings


# ----------------------------- CLI ---------------------------------------

def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description="VOICEVOX 音声合成（栗田まろん）")
    parser.add_argument("--input", type=str, default=None, help="script.json のパス（full_script を読む）")
    parser.add_argument("--text", type=str, default=None, help="合成する文字列を直接指定")
    parser.add_argument("--output", type=str, default=None, help="出力 WAV パス")
    parser.add_argument("--speaker", type=str, default=DEFAULT_SPEAKER_NAME, help="スピーカー名")
    parser.add_argument("--style", type=str, default=DEFAULT_STYLE_NAME, help="スタイル名（例: 喜び）")
    parser.add_argument(
        "--no-flatten",
        action="store_true",
        help="語尾下げ調整を無効化（VOICEVOX 既定の抑揚で合成）",
    )
    args = parser.parse_args(argv)

    if not check_engine():
        print(
            "[ERROR] VOICEVOX が起動していません。\n"
            f"  期待するエンドポイント: {VOICEVOX_HOST}\n"
            "  VOICEVOX を起動してから再実行してください。",
            file=sys.stderr,
        )
        return 1

    try:
        speaker_id = find_speaker_id(args.speaker, args.style)
    except VOICEVOXError as e:
        print(f"[ERROR] {e}", file=sys.stderr)
        return 1
    style_label = f" / {args.style}" if args.style else ""
    print(f"使用スピーカー: {args.speaker}{style_label} (style_id={speaker_id})")

    # 入力テキスト & 出力パスの決定
    if args.input:
        in_path = Path(args.input)
        if not in_path.exists():
            print(f"[ERROR] {in_path} がありません", file=sys.stderr)
            return 1
        script = json.loads(in_path.read_text(encoding="utf-8"))
        text = script.get("full_script", "").strip()
        if not text:
            print("[ERROR] script.json に full_script がありません", file=sys.stderr)
            return 1
        out_path = Path(args.output) if args.output else (in_path.parent / "voice.wav")
    elif args.text:
        text = args.text
        out_path = Path(args.output) if args.output else Path("voice.wav")
    else:
        print("[ERROR] --input か --text を指定してください", file=sys.stderr)
        return 1

    flatten = not args.no_flatten
    print(f"合成テキスト ({len(text)} 字): {text[:60]}...")
    print(f"語尾下げ調整: {'ON (B案)' if flatten else 'OFF'}")
    print("合成中...")
    try:
        wav_bytes, timings = synthesize_full(text, speaker_id, flatten_endings=flatten)
    except Exception as e:
        print(f"[ERROR] 合成失敗: {e}", file=sys.stderr)
        return 1

    out_path.parent.mkdir(parents=True, exist_ok=True)
    out_path.write_bytes(wav_bytes)

    # 字幕タイミングを保存（reel_creator が読み取って同期に使う）
    timings_path = out_path.parent / "timings.json"
    timings_path.write_text(
        json.dumps(timings, ensure_ascii=False, indent=2),
        encoding="utf-8",
    )

    with wave.open(io.BytesIO(wav_bytes), "rb") as w:
        duration_sec = w.getnframes() / w.getframerate()
    size_kb = len(wav_bytes) / 1024

    print(f"音声出力 : {out_path}  ({size_kb:.1f} KB, {duration_sec:.2f} 秒)")
    print(f"タイミング: {timings_path}  ({len(timings)} 文)")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: reel-bot/src/voicevoxClient.js
// ② VOICEVOX 音声合成（青山龍星 / 喜び）
// - 文を 。！？ で分割して個別合成
// - 文末アクセント句の末尾モーラ pitch を下げて語尾上がりを抑制（B案）
// - 文間に無音を挟んで連結し、各文の start/end（timings）を返す

import fs from "fs";
import path from "path";
import { env, VOICE, OUTPUT_REELS } from "./config.js";
import { parseWav, buildWav, silencePcm, pcmDurationSec } from "./wav.js";
import { CUSTOM_READINGS } from "./voicevoxDict.js";

const HOST = env.VOICEVOX_HOST;

export async function checkEngine() {
  try {
    const res = await fetch(`${HOST}/version`, { signal: AbortSignal.timeout(3000) });
    return res.ok;
  } catch {
    return false;
  }
}

// ユーザー辞書の登録（読み間違い矯正）。未登録の surface だけ追加する。
export async function registerUserDict(readings = CUSTOM_READINGS) {
  if (!readings?.length) return;
  let existingSurfaces = new Set();
  try {
    const res = await fetch(`${HOST}/user_dict`, { signal: AbortSignal.timeout(10000) });
    if (res.ok) {
      const dict = await res.json(); // { uuid: { surface, ... } }
      existingSurfaces = new Set(Object.values(dict).map((w) => w.surface));
    }
  } catch {
    /* 取得失敗時はそのまま登録を試みる */
  }
  for (const r of readings) {
    if (existingSurfaces.has(r.surface)) continue;
    const url =
      `${HOST}/user_dict_word?surface=${encodeURIComponent(r.surface)}` +
      `&pronunciation=${encodeURIComponent(r.pronunciation)}` +
      `&accent_type=${r.accent_type ?? 0}` +
      `&priority=${r.priority ?? 10}`; // 高優先度で文中の誤分割を防ぐ
    try {
      const res = await fetch(url, { method: "POST", signal: AbortSignal.timeout(10000) });
      if (!res.ok) console.log(`  (note) 辞書登録失敗 ${r.surface}: HTTP ${res.status}`);
    } catch (e) {
      console.log(`  (note) 辞書登録エラー ${r.surface}: ${e.message}`);
    }
  }
}

export async function findSpeakerId(speaker = VOICE.speaker, style = VOICE.style) {
  const res = await fetch(`${HOST}/speakers`, { signal: AbortSignal.timeout(30000) });
  if (!res.ok) throw new Error(`/speakers 失敗: HTTP ${res.status}`);
  const speakers = await res.json();
  const sp = speakers.find((s) => s.name === speaker);
  if (!sp) {
    const names = speakers.map((s) => s.name).join(", ");
    throw new Error(`スピーカー '${speaker}' が見つかりません。利用可能: ${names}`);
  }
  if (style) {
    const st = sp.styles.find((x) => x.name === style);
    if (!st) {
      const styles = sp.styles.map((x) => x.name).join(", ");
      throw new Error(`'${speaker}' にスタイル '${style}' がありません。利用可能: ${styles}`);
    }
    return st.id;
  }
  return sp.styles[0].id;
}

// 文末アクセント句の末尾モーラ pitch を下げる（B案）
function flattenPhraseEndings(query) {
  const phrases = query.accent_phrases || [];
  if (phrases.length === 0) return query;
  const lastPhrase = phrases[phrases.length - 1];
  const moras = lastPhrase.moras || [];
  if (moras.length === 0) return query;

  const last = moras[moras.length - 1];
  if (last.pitch > 0) last.pitch = Math.round(last.pitch * VOICE.lastMoraFactor * 1000) / 1000;
  if (moras.length >= 2) {
    const second = moras[moras.length - 2];
    if (second.pitch > 0)
      second.pitch = Math.round(second.pitch * VOICE.secondLastMoraFactor * 1000) / 1000;
  }
  return query;
}

export async function synthesizeOne(text, speakerId, { flatten = true } = {}) {
  // 1. audio_query
  const q = await fetch(
    `${HOST}/audio_query?speaker=${speakerId}&text=${encodeURIComponent(text)}`,
    { method: "POST", signal: AbortSignal.timeout(30000) }
  );
  if (!q.ok) throw new Error(`audio_query 失敗: HTTP ${q.status}`);
  const query = await q.json();
  query.speedScale = VOICE.speed;
  query.pitchScale = VOICE.pitch;
  query.intonationScale = VOICE.intonation;
  if (flatten) flattenPhraseEndings(query);

  // 2. synthesis
  const s = await fetch(`${HOST}/synthesis?speaker=${speakerId}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(query),
    signal: AbortSignal.timeout(120000),
  });
  if (!s.ok) throw new Error(`synthesis 失敗: HTTP ${s.status}`);
  return Buffer.from(await s.arrayBuffer());
}

export function splitSentences(text) {
  return text
    .split(/(?<=[。！？])/)
    .map((s) => s.trim())
    .filter(Boolean);
}

// 全文を合成 → {wavBuffer, timings}
// displayScript: 字幕・フックに表示する原文（漢字混じり）
// opts.voiceScript: 音声合成に使う変換後テキスト。省略時は displayScript と同一
//   - 文区切り（。！？）の数は displayScript と一致している必要がある
//   - timings.text には displayScript の文がそのまま入る（字幕は表示テキストで描画）
export async function synthesizeFull(displayScript, speakerId, { flatten = true, voiceScript = null } = {}) {
  const displaySentences = splitSentences(displayScript);
  if (displaySentences.length === 0) throw new Error("台本が空です");
  const voiceSentences = voiceScript ? splitSentences(voiceScript) : displaySentences;
  if (voiceSentences.length !== displaySentences.length) {
    throw new Error(
      `表示文と音声文の数が一致しません: display=${displaySentences.length} voice=${voiceSentences.length}。` +
      `voiceText の変換で 。！？ の数が変わっていないか確認してください。`
    );
  }

  await registerUserDict(); // 合成前に読み矯正辞書を登録

  let fmt = null;
  const pcmChunks = [];
  const durations = [];

  for (const sent of voiceSentences) {
    const wav = await synthesizeOne(sent, speakerId, { flatten });
    const { fmt: f, data } = parseWav(wav);
    if (!fmt) fmt = f;
    pcmChunks.push(data);
    durations.push(pcmDurationSec(f, data));
  }

  const silence = silencePcm(fmt, VOICE.silenceSec);

  const parts = [];
  const timings = [];
  let cursor = 0;
  for (let i = 0; i < displaySentences.length; i++) {
    timings.push({
      text: displaySentences[i], // 字幕には表示用テキスト（漢字）を残す
      start: Math.round(cursor * 1000) / 1000,
      end: Math.round((cursor + durations[i]) * 1000) / 1000,
    });
    parts.push(pcmChunks[i]);
    cursor += durations[i];
    if (i < displaySentences.length - 1) {
      parts.push(silence);
      cursor += VOICE.silenceSec;
    }
  }

  const wavBuffer = buildWav(fmt, Buffer.concat(parts));
  return { wavBuffer, timings };
}

// ---------- CLI ----------
function parseArgs(argv) {
  const a = {};
  for (let i = 0; i < argv.length; i++) {
    if (argv[i] === "--input") a.input = argv[++i];
    else if (argv[i] === "--text") a.text = argv[++i];
    else if (argv[i] === "--output") a.output = argv[++i];
    else if (argv[i] === "--no-flatten") a.noFlatten = true;
  }
  return a;
}

async function main() {
  const args = parseArgs(process.argv.slice(2));

  if (!(await checkEngine())) {
    console.error(
      `[ERROR] VOICEVOX が起動していません（${HOST}）。VOICEVOX を起動してから再実行してください。`
    );
    process.exit(1);
  }

  let speakerId;
  try {
    speakerId = await findSpeakerId();
  } catch (e) {
    console.error(`[ERROR] ${e.message}`);
    process.exit(1);
  }
  console.log(`使用スピーカー: ${VOICE.speaker} / ${VOICE.style} (style_id=${speakerId})`);

  let text, outPath;
  if (args.input) {
    const script = JSON.parse(fs.readFileSync(args.input, "utf-8"));
    text = (script.full_script || "").trim();
    if (!text) {
      console.error("[ERROR] script.json に full_script がありません");
      process.exit(1);
    }
    outPath = args.output || path.join(path.dirname(args.input), "voice.wav");
  } else if (args.text) {
    text = args.text;
    outPath = args.output || "voice.wav";
  } else {
    console.error("[ERROR] --input か --text を指定してください");
    process.exit(1);
  }

  console.log(`語尾下げ調整: ${args.noFlatten ? "OFF" : "ON (B案)"}`);
  console.log(`合成テキスト(${text.length}字): ${text.slice(0, 60)}...`);
  console.log("合成中...");

  const { wavBuffer, timings } = await synthesizeFull(text, speakerId, { flatten: !args.noFlatten });
  fs.mkdirSync(path.dirname(outPath), { recursive: true });
  fs.writeFileSync(outPath, wavBuffer);

  const timingsPath = path.join(path.dirname(outPath), "timings.json");
  fs.writeFileSync(timingsPath, JSON.stringify(timings, null, 2), "utf-8");

  const { fmt, data } = parseWav(wavBuffer);
  const dur = pcmDurationSec(fmt, data);
  console.log(`音声出力 : ${outPath} (${(wavBuffer.length / 1024).toFixed(1)} KB, ${dur.toFixed(2)} 秒)`);
  console.log(`タイミング: ${timingsPath} (${timings.length} 文)`);
}

if (import.meta.url === `file://${process.argv[1]}` || process.argv[1]?.endsWith("voicevoxClient.js")) {
  main().catch((e) => {
    console.error(`[ERROR] ${e.message}`);
    process.exit(1);
  });
}
